### PR TITLE
Multinode on MPS

### LINKED ISSUE #15713: Multinode standalone not working on mps
### Bug description

When running on  MPS capable devices, the multinode component sets accelerator="auto", devices="auto" and strategy="ddp" which will result in MPSAccelerator. 
Since distributed on MPS is not supported, it has to be run on CPU for all apple silicon platforms instead.

### How to reproduce the bug

```python
# app.py
import lightning as L
from lightning.app.components import LightningTrainerMultiNode
from lightning.pytorch.demos.boring_classes import BoringModel


class LightningTrainerDistributed(L.LightningWork):
    @staticmethod
    def run():
        model = BoringModel()
        trainer = L.Trainer(max_epochs=10, strategy="ddp") # same with any combination of strategy, accelerator and devices as this will be handled by the Multinodecomponent
        trainer.fit(model)

# 8 GPU: (2 nodes of 4 x v100)
component = LightningTrainerMultiNode(
    LightningTrainerDistributed,
    num_nodes=4,
    cloud_compute=L.CloudCompute("gpu-fast-multi"), # 4 x v100
)
app = L.LightningApp(component)
 

on apple silicon devices
```

### Error messages and logs

```
-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/torch/multiprocessing/spawn.py", line 69, in _wrap
    fn(i, *args)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/app/components/multi_node/pytorch_spawn.py", line 58, in dispatch_run
    cls.run(local_rank, unwrap(work.run), *args, **kwargs)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/app/components/multi_node/trainer.py", line 69, in run
    work_run()
  File "/Users/nohaalon/Desktop/bcn/app2.py", line 12, in run
    trainer.fit(model)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 578, in fit
    call._call_and_handle_interrupt(
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/pytorch/trainer/call.py", line 38, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 620, in _fit_impl
    self._run(model, ckpt_path=self.ckpt_path)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/pytorch/trainer/trainer.py", line 1038, in _run
    self.strategy.setup(self)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/pytorch/strategies/ddp.py", line 179, in setup
    self.configure_ddp()
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/pytorch/strategies/ddp.py", line 295, in configure_ddp
    self.model = self._setup_model(LightningDistributedModule(self.model))
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/lightning/pytorch/strategies/ddp.py", line 195, in _setup_model
    return DistributedDataParallel(module=model, device_ids=device_ids, **self._ddp_kwargs)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/torch/nn/parallel/distributed.py", line 655, in __init__
    _verify_param_shape_across_processes(self.process_group, parameters)
  File "/Users/nohaalon/opt/miniconda3/envs/stag-34/lib/python3.10/site-packages/torch/distributed/utils.py", line 112, in _verify_param_shape_across_processes
    return dist._verify_params_across_processes(process_group, tensors, logger)
RuntimeError: ProcessGroupGloo::allgather: unsupported device type mps

```

### Environment

_No response_

### More info

Cannot handle for plain PyTorch, need to handle for Lite and Trainer MultiNode.

cc @tchaton @nohalon

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Fixes #15713 

Distributed Training is not supported on MPS devices -> catches this for Lite and PL Multinode components, nothing we can do on raw pytorch though.

Also adds unittests for lite and pl multinode executors beyond the changes on this PR.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/components/multi_node/pytorch_spawn.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def run(
         elif world_size > 1:
             raise Exception("Torch distributed should be available.")
 
-        work_run(world_size, node_rank, global_rank, local_rank)
+        return work_run(world_size, node_rank, global_rank, local_rank)
```
Comment (tchaton) [inline src/lightning_app/components/multi_node/pytorch_spawn.py:91]: This won't work. We would need to use queues to bring back the work state and the return value. 
Do you want to try to address it ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning_app/components/multi_node/pytorch_spawn.py (hunk 1 above)
Comment (justusschock) [reply src/lightning_app/components/multi_node/pytorch_spawn.py:91]: that's only for standalone testing here. I am aware that with mp.spawn this won't work that way.

**Review by awaelchli — APPROVED** [review]
Thanks for adding the missing tests! Should have been there from the beginning

**Comment by justusschock** [comment]
@Borda Yes it is and this is why it is skipped for non-mps.

The other tests are missing tests added to ensure I break nothing

File: src/lightning_app/components/multi_node/lite.py [hunk 2]
```diff
@@ -52,7 +54,18 @@ def run(
...
+                old_acc_value = kwargs.get("accelerator", "auto")
+                kwargs["accelerator"] = "cpu"
+
+                if old_acc_value != kwargs["accelerator"]:
+                    warnings.warn(
+                        "Forcing accelerator=cpu as other accelerators (specifically MPS) are not supported "
+                        "by PyTorch for distributed training on mps capable devices"
+                    )
+            else:
+                kwargs["accelerator"] = "auto"
```
[the remark this answers is not in the document]
Comment (justusschock) [reply src/lightning_app/components/multi_node/lite.py:68]: i find this more explicit and easier readible

File: .github/workflows/ci-app-tests.yml [hunk 3]
```diff
@@ -105,6 +105,7 @@ jobs:
       - name: Adjust examples
         if: ${{ matrix.pkg-name != 'lightning' }}
         run: |
+          python .actions/assistant.py copy_replace_imports --source_dir="./tests" --source_import="lightning.app,lightning" --target_import="lightning_app,lightning_app"
```
Comment (carmocca) [inline .github/workflows/ci-app-tests.yml:108]: Is it expected that tests files use the new import style? Wouldn't it be better to update those usages?

File: .github/workflows/ci-app-tests.yml [hunk 4]
```diff
@@ -72,23 +72,23 @@ jobs:
...
       - name: Switch PyTorch URL
         run: python -c "print('TORCH_URL=https://download.pytorch.org/whl/' + str('test/cpu/torch_test.html' if '${{matrix.release}}' == 'pre' else 'cpu/torch_stable.html'))" >> $GITHUB_ENV
 
-      - name: Install package
+      - name: Install package & dependencies
         env:
           PACKAGE_NAME: ${{ matrix.pkg-name }}
         run: |
-          pip install -e . pytest --upgrade --find-links ${TORCH_URL}
+          pip install -e . -r requirements/app/devel.txt -U -q --find-links ${TORCH_URL}
```
Comment (carmocca) [inline .github/workflows/ci-app-tests.yml:85]:
```suggestion
          pip install -e .[dev] -U -q --find-links ${TORCH_URL}
```

File: src/lightning_app/CHANGELOG.md [hunk 5]
```diff
@@ -52,12 +55,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 - Fixed CloudRuntime works collection with structures and accelerated multi node startup time ([#15650](https://github.com/Lightning-AI/lightning/pull/15650))
 - Fixed catimage import ([#15712](https://github.com/Lightning-AI/lightning/pull/15712))
-- Parse all lines in app file looking for shebangs to run commands ([#15714](https://github.com/Lightning-AI/lightning/pull/15714))
-
+- Parse all lines in app file looking for shebangs to run commands ([#15714](https://github.com/Lightning-AI/lightning/pull/15714)
 - Fixed setting property to the LightningFlow ([#15750](https://github.com/Lightning-AI/lightning/pull/15750))
 
 
 
+
```
Comment (carmocca) [inline src/lightning_app/CHANGELOG.md:63]:
```suggestion
```

File: src/lightning_app/components/multi_node/lite.py [hunk 6]
```diff
@@ -52,7 +54,18 @@ def run(
...
-            kwargs["accelerator"] = "auto"
+
+            if MPSAccelerator.is_available():
+                old_acc_value = kwargs.get("accelerator", "auto")
+                kwargs["accelerator"] = "cpu"
+
+                if old_acc_value != kwargs["accelerator"]:
+                    warnings.warn(
+                        "Forcing accelerator=cpu as other accelerators (specifically MPS) are not supported "
+                        "by PyTorch for distributed training on mps capable devices"
```
Comment (carmocca) [inline src/lightning_app/components/multi_node/lite.py:65]: This message is a bit confusing. Why not

```suggestion
                        "Forcing accelerator=cpu as MPS does not support distributed training"
```

File: src/lightning_app/components/multi_node/trainer.py [hunk 7]
```diff
@@ -50,7 +52,15 @@ def run(
         def pre_fn(trainer, *args, **kwargs):
             kwargs["devices"] = nprocs
             kwargs["num_nodes"] = num_nodes
-            kwargs["accelerator"] = "auto"
+            if MPSAccelerator.is_available():
```
Comment (carmocca) [inline src/lightning_app/components/multi_node/trainer.py:55]: Don't we need the `else` here as above?

File: src/lightning_app/components/multi_node/trainer.py [hunk 8]
```diff
@@ -50,7 +52,15 @@ def run(
...
+                old_acc_value = kwargs.get("accelerator", "auto")
+                kwargs["accelerator"] = "cpu"
+
+                if old_acc_value != kwargs["accelerator"]:
+                    warnings.warn(
+                        "Forcing accelerator=cpu as other accelerators (specifically MPS) are not supported "
```
Comment (carmocca) [inline src/lightning_app/components/multi_node/trainer.py:61]: Same suggestion

File: .github/workflows/ci-app-tests.yml (hunk 3 above)
Comment (justusschock) [reply .github/workflows/ci-app-tests.yml]: nope, that was my mistake. Has been updated

File: src/lightning_app/components/multi_node/trainer.py (hunk 7 above)
Comment (justusschock) [reply src/lightning_app/components/multi_node/trainer.py:55]: yep missed that, thanks

File: tests/tests_app/components/multi_node/test_lite.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
+import os
+from copy import deepcopy
+from functools import partial
+from unittest import mock
+
+import pytest
+from lightning_utilities.core.imports import module_available
+from tests_app.helpers.utils import no_warning_call
+
+import lightning_lite as ll
```
Comment (awaelchli) [inline tests/tests_app/components/multi_node/test_lite.py:10]: Try to change it to lightning.lite. The tracer only patches these imports (could be considered a bug)

Comment (awaelchli) [inline tests/tests_app/components/multi_node/test_lite.py:10]:
```suggestion
import lightning.lite as ll
```

File: src/lightning_app/components/multi_node/trainer.py [hunk 10]
```diff
@@ -50,24 +59,37 @@ def run(
...
             kwargs["num_nodes"] = num_nodes
-            kwargs["accelerator"] = "auto"
+            if any(x.is_available() for x in mps_accelerators):
+                old_acc_value = kwargs.get("accelerator", "auto")
+                kwargs["accelerator"] = "cpu"
+
+                if old_acc_value != kwargs["accelerator"]:
+                    warnings.warn(
+                        "Forcing accelerator=cpu as other accelerators (specifically MPS) are not supported "
+                        "by PyTorch for distributed training on mps capable devices"
```
Comment (Borda) [inline src/lightning_app/components/multi_node/trainer.py:69]:
```suggestion
                        "Forcing `accelerator=cpu` as MPS does not support distributed training."
```

**Comment by Borda** [comment]
@justusschock, I have pretty much just set to run the test for `lightning` only not for Lite/App/Pytorch... I hope it is fine, mind checking it :otter:

File: src/lightning_app/components/multi_node/lite.py [hunk 11]
```diff
@@ -52,23 +62,36 @@ def run(
         def pre_fn(lite, *args, **kwargs):
             kwargs["devices"] = nprocs
             kwargs["num_nodes"] = num_nodes
-            kwargs["accelerator"] = "auto"
+
+            if any(x.is_available() for x in mps_accelerators):
```
Comment (Borda) [inline src/lightning_app/components/multi_node/lite.py:66]:
```suggestion
            if any(acc.is_available() for acc in mps_accelerators):
```

File: src/lightning_app/components/multi_node/trainer.py [hunk 12]
```diff
@@ -50,24 +59,34 @@ def run(
         def pre_fn(trainer, *args, **kwargs):
             kwargs["devices"] = nprocs
             kwargs["num_nodes"] = num_nodes
-            kwargs["accelerator"] = "auto"
+            if any(x.is_available() for x in mps_accelerators):
```
Comment (Borda) [inline src/lightning_app/components/multi_node/trainer.py:62]:
```suggestion
            if any(acc.is_available() for acc in mps_accelerators):
```

File: tests/tests_app/components/multi_node/test_trainer.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+@pytest.mark.parametrize(
+    "args_given,args_expected",
+    [
+        (
+            {
+                "devices": 1,
+                "num_nodes": 1,
+                "accelerator": "gpu",
+            },
+            {"devices": 8, "num_nodes": 7, "accelerator": "auto"},
```
Comment (Borda) [inline tests/tests_app/components/multi_node/test_trainer.py:70]:
```suggestion
            {
                "devices": 1,
                "num_nodes": 1,
                "accelerator": "gpu"
            },
            {"devices": 8, "num_nodes": 7, "accelerator": "auto"}
```

**Review by Borda — APPROVED** [review]

**Pull request merged**